[RFC] Mqtt transport: Redesign the service to cover more usecases.

davidgraeff opened the pull request:
Fixes #3695 
This is a request for comments PR, discussion takes places in the linked issue.

* This will keep the 'old' MqttService API untouched for full backwards compatibility.
* MqttBrokerConnection is exported now.
* Improve comments all over the code.

MqttService:
* Allows to add / remove / enumerate broker connections, additionally to the existing textual configuration.
* A new interface allows listeners to be notified of added / removed broker connections.
* All proxy methods are deprecated now. The MqttBrokerConnection can be accessed directly.
* Add unit test

MqttWillAndTestament:
* Introduce an isValid() method and return false if the build method fromString is not provided with all necessary properties.

MqttSenderChannel:
* Is depreacted now. MqttBrokerConnection.publish() should be used instead.
* The reason: The interface does not allow to implement an async behaviour, because the publisher can not be informed of the success/failure of the operation.

MqttMessageConsumer:
* is deprecated now and replaced by a slimer MqttMessageSubscriber interface without a forced setTopic and event method.  

MqttBrokerConnection:
* Reconnection is no longer handled via a fixed 60s interval timer, but fully customizable.
  The default implementation (PeriodicReconnectPolicy) resembles the old behaviour.
* A SSL broker connection no longer uses a non-configurable "trust all and everything" SSL context provider, but an interface that have to provide the SSL context provider. A default implementation, again, resembles the old behaviour.
* Remove 'async' configuration option. All communication aspects (connect/disconnect/publish/subscribe) are handled in an async fashion now.
* Add unit test

Signed-off-by: David Gräff <david.graeff@web.de>

maggu2810 commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttWillAndTestament.java:
```diff
@@ -45,39 +45,39 @@ public static MqttWillAndTestament fromString(String string) {
-    public String getTopic() {
-        return topic;
+    private boolean isValid() {
+        return !StringUtils.isBlank(topic) && payload.length > 0;
```
An 0 byte will message is valid
http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Toc385349244

maggu2810 replied:
It's been some time since I read the spec and wrote a MQTT broker, so excuse me if it is not correct, but if IIRC:
If the retained flag is used for MQTT messages, a null message is used to "un"-retain a topic again.
So using this approach to signal a disconnect (remove a retained topic) is a valid approach.

Miss I something?

davidgraeff replied:
True. isValid() should solely check the topic.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java:
```diff
@@ -29,211 +34,315 @@
+
+        for (Map<String, Object> brokerConfig : brokerConfigs.values()) {
+            try {
+                addBrokerConnection(brokerConfig).start();
```
Already available connections to the named broker will be returned but updated config will not be taken into account. Should an updated configuration restart existing connections with updated parameters?

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java:
```diff
@@ -29,211 +34,315 @@
+     */
+    public void addBrokerConnection(MqttBrokerConnection connection) {
+        synchronized (brokerConnections) {
+            brokerConnections.put(connection.getName(), connection);
```
Is it guaranteed that externally created BrokerConnections also have an all lower case name? Otherwise you will not be able to get it back from the MqttService (see line 144).

davidgraeff replied:
It is not guaranteed, I will force lowercase names here and add a boolean return value (true==successfully added, false==already existing).

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java:
```diff
@@ -29,211 +34,315 @@
-            if (conn == null) {
-                conn = new MqttBrokerConnection(brokerName);
-                brokerConnections.put(brokerName.toLowerCase(), conn);
+            brokerConnections.remove(connection.getName(), connection);
```
This again will only work for external connections if the name is not all lower case.
In addition you may check the return value and call the observers only if the removal succeeded.

davidgraeff replied:
I agree

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java:
```diff
@@ -29,211 +34,315 @@
+            if (connection == null) {
+                return;
+            }
+            brokerConnections.remove(brokerName);
```
here again `brokerName` may not be equal to `brokerName.toLowerCase()` in which case the observers get notified but the connection is still in the map. You may call `brokerConnections.remove(brokerName.toLowerCase(), connection)` here and again check the return value. If the remove fails you may log.debug to give a hint what happened.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java:
```diff
@@ -29,211 +34,315 @@
+    @Deprecated
     public void registerMessageConsumer(String brokerName, MqttMessageConsumer mqttMessageConsumer) {
-
         mqttMessageConsumer.setEventPublisher(eventPublisher);
```
not part of your PR and deprecated but I guess this should be done in line 283 only when the consumer was added to the connection.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/MqttSenderChannelImpl.java:
```diff
@@ -0,0 +1,32 @@
+package org.eclipse.smarthome.io.transport.mqtt.internal;
```
missing license text

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/reconnect/AbstractReconnectPolicy.java:
```diff
@@ -0,0 +1,25 @@
+package org.eclipse.smarthome.io.transport.mqtt.reconnect;
```
missing license text

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/reconnect/PeriodicReconnectPolicy.java:
```diff
@@ -0,0 +1,58 @@
+package org.eclipse.smarthome.io.transport.mqtt.reconnect;
```
missing license text

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/sslcontext/AcceptAllCertifcatesSSLContext.java:
```diff
@@ -0,0 +1,45 @@
+package org.eclipse.smarthome.io.transport.sslcontext;
```
missing license text

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/sslcontext/SSLContextProvider.java:
```diff
@@ -0,0 +1,8 @@
+package org.eclipse.smarthome.io.transport.sslcontext;
```
missing license text

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java:
```diff
@@ -29,211 +34,315 @@
+     */
+    public MqttBrokerConnection addBrokerConnection(Map<String, Object> brokerConnectionConfig)
+            throws ConfigurationException, MqttException {
+        String name = brokerConnectionConfig.get("name").toString();
```
`"name"` should be a constant.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java:
```diff
@@ -29,211 +34,315 @@
+            if (brokerConfig == null) {
+                brokerConfig = new TreeMap<>();
+                configPerBroker.put(brokername, brokerConfig);
+                brokerConfig.put("name", brokername);
```
`"name"` should be a constant.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java:
```diff
@@ -29,211 +34,315 @@
+            throws ConfigurationException, MqttException {
+        String name = brokerConnectionConfig.get("name").toString();
+        if (name == null) {
+            throw new ConfigurationException("MQTT Broker property '{}' should have the format 'broker.propertykey'");
```
there is no value for `{}` here. You can name the `name` property right away here.

davidgraeff commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java:
```diff
@@ -29,211 +34,315 @@
+     * @return A 'list' of broker configurations as key-value maps. A configuration map at least contains a "name".
+     */
+    public Map<String, Map<String, Object>> extractBrokerConfigurations(Dictionary<String, ?> properties) {
+        Map<String, Map<String, Object>> configPerBroker = new TreeMap<String, Map<String, Object>>();
```
I prefer TreeMaps over HashMaps if I expect only a few entries and if only insert and no remove operation are performed (in this case in average 3 configuration entries per broker are added) to reduce the memory footprint. But if ESH Guidelines prefer HashMaps, I'll change it.

htreu replied:
I dont think this is a ESH guideline and I also guess the memory difference is not that big. What do you think @sjka?

sjka replied:
No, I don't think we have a strict written guideline about it.

Nevertheless, it is common practice in this project to use sorted collections such as TreeSet, TreeMap etc. only if the order of elements has a relevant meaning. I don't think this is the case here, is it?

davidgraeff replied:
The order is irrelevant. I will change it then. I was just concerned of the in average 13 unused entries in a hash map (with an initial default capacity of 16). But because we store string references only it might be just a few bytes per entry anyway.

davidgraeff commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java:
```diff
@@ -29,211 +34,315 @@
+    public void updated(Dictionary<String, ?> properties) {
+        // load broker configurations from configuration file
+        if (properties == null || properties.isEmpty()) {
+            return;
```
Good point, to be sure I will disconnect and discard all existing, textual configured broker connections

davidgraeff commented:
I addressed all issues except the unit tests, but cannot build right now. My eclipse is not able to find the  org.eclipse.paho.client.mqttv3 library anymore in the target configuration since the last update.

davidgraeff commented:
I have added unit tests as well for MqttService and MqttBrokerConnection (except start/close).

davidgraeff commented:
I've got one question: Should we keep the deprecated methods and classes? If there are no consumers of this API and ESH hasn't reached 1.0 it might be a good idea to remove them.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java:
```diff
@@ -25,215 +31,375 @@
+     * @throws ConfigurationException Most likely your provided name and url are invalid.
+     * @throws MqttException
+     */
+    public MqttBrokerConnection addBrokerConnection(Map<String, Object> brokerConnectionConfig)
```
What other types of values instead of `String` do you expect here? I would suggest to expect a Map<String, String> which would result in property retrieval like

```
String brokerURL = brokerConnectionConfig.get("url");
if (StringUtils.isBlank(brokerURL) {
    throw new ConfigurationException("MQTT Broker property 'url' is not provided");
}
```

Right now passing in a blank String would satisfy all the checks but will fail on connecting to the broker.
On another note this would also make the check in line 66 obsolete.

davidgraeff replied:
I agree and adapted the code

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/MqttService.java:
```diff
@@ -25,215 +31,375 @@
             String[] subkeys = key.split("\\.");
-            if (subkeys.length != 2) {
-                logger.debug("MQTT Broker property '{}' should have the format 'broker.propertykey'", key);
+            if (subkeys.length != 2 || StringUtils.isBlank(value)) {
```
Are blank values a problem? Only `name` and `url` are required. Other properties might be blank. Its okay to check it here but with a brokerConfig map being <String, String> and a public API allowing to put even blank Strings (see my comment on line 203) this check is obsolete.

davidgraeff replied:
No they are not a problem therefore we don't fail here, but log the unexpected name.key=no_value triple as debug log level. Improper used "key=" lines should be commented out instead of keeping the parser busy, I reckon.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttBrokerConnectionTests.java:
```diff
@@ -0,0 +1,185 @@
+        // Test an invalid URL
+        try {
+            @SuppressWarnings("unused")
+            MqttBrokerConnection ignored = new MqttBrokerConnection("name", "unsupported://123.123.123.123", false);
```
I guess this is a matter of taste: I would prefer separate tests for these assertions. You may then shorten the test to something like

```
@Test(expected = ConfigurationException.class)
public void unsupportedProtocolShouldThrowException() throws Exception {
    new MqttBrokerConnection("name", "unsupported://123.123.123.123", false);
}
```

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttBrokerConnectionTests.java:
```diff
@@ -0,0 +1,185 @@
+        MqttBrokerConnection a;
+        try {
+            a = new MqttBrokerConnection(name, url, false);
+        } catch (ConfigurationException e) {
```
You may declare the test to `throws Exception` and just call the constructor here. This would clean up the test and it will fail anyway if the exception occurs.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttBrokerConnectionTests.java:
```diff
@@ -0,0 +1,185 @@
+            }
+        };
+        try {
+            a.addConsumer(subscriber);
```
same as in line 71/72. Just call `addConsumer` and let the test fail due to the anyhow thrown exception.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttBrokerConnectionTests.java:
```diff
@@ -0,0 +1,185 @@
+        };
+        a.addConnectionObserver(connectionObserver);
+        // Adding a connection observer should immediately call its setConnected() method.
+        assertFalse(isConnected);
```
Just as an alternative path using a mock:

```
MqttConnectionObserver connectionObserver = mock(MqttConnectionObserver.class);
a.addConnectionObserver(connectionObserver);
verify(connectionObserver).setConnected(false);
```

wdyt?

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttBrokerConnectionTests.java:
```diff
@@ -0,0 +1,185 @@
+            fail();
+            return;
+        }
+        assertTrue("URL getter", a.getUrl().equals(url));
```
Please use `assertEquals(url, a.getUrl());` or even better `assertThat(a.getUrl(), is(url));` here. Please have a look at the other assertions where semantically more appropriate methods are available.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttBrokerConnectionTests.java:
```diff
@@ -0,0 +1,185 @@
+        assertNull(a.getLastWill());
+        assertNull(MqttWillAndTestament.fromString(""));
+        a.setLastWill(MqttWillAndTestament.fromString("topic:message:1:true"));
+        assertNotNull(a.getLastWill());
```
this is obsolete since an NPE in the next line would also fail the test.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttServiceTests.java:
```diff
@@ -0,0 +1,117 @@
+            return;
+        }
+        assertTrue(service.addBrokerConnection(connection));
+        assertTrue(brokerExisting);
```
Just like in the `MqttBrokerConnectionTests` this would be a little shorter with a mock `MqttBrokersObserver` and `org.mockito.Mockito.vrify`.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttServiceTests.java:
```diff
@@ -0,0 +1,117 @@
+        assertTrue(brokerExisting);
+
+        service.removeBrokerConnection(connection);
+        assertFalse(brokerExisting);
```
same as in line 59.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttServiceTests.java:
```diff
@@ -0,0 +1,117 @@
+        properties.put("bam.name", "brokername");
+        properties.put("bam.url", "tcp://123.123.123.123");
+        Map<String, Map<String, Object>> map = service.extractBrokerConfigurations(properties);
+        assertNotNull(map);
```
obsolete, will fail with NPE in the next line.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttServiceTests.java:
```diff
@@ -0,0 +1,117 @@
+        assertNotNull(map);
+        assertEquals(map.size(), 1);
+        Map<String, Object> data = map.get("bam");
+        assertNotNull(data);
```
same as line 77.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttServiceTests.java:
```diff
@@ -0,0 +1,117 @@
+        assertTrue("brokername".equals(data.get("name")));
+        assertTrue("tcp://123.123.123.123".equals(data.get("url")));
+
+        assertEquals(service.getAllBrokerConnections().size(), 0);
```
good. even better: `assertThat(service.getAllBrokerConnections(), hasSize(0));` with `hasSize` being an static import of `IsCollectionWithSize.hasSize`. 
All in all: matter of taste.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttServiceTests.java:
```diff
@@ -0,0 +1,117 @@
+        assertEquals(service.getAllBrokerConnections().size(), 0);
+        try {
+            service.addBrokerConnection(data);
+        } catch (ConfigurationException | MqttException e) {
```
same as in the `MqttBrokerConnectionTests`. You dont want to test the Exception here. Just let it fail with its own Exception.

davidgraeff commented:
Thanks for the test suite review, your suggestions are quite helpful and I will adapt the tests accordingly. In the Openhab world, writing tests with eclipse doesn't work right now because of https://github.com/openhab/openhab-distro/issues/230 so I'm a little inexperienced in this area.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttServiceTests.java:
```diff
@@ -19,75 +21,60 @@
-    boolean brokerExisting = false;
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
```
You can remove `MockitoAnnotations.initMocks(this)` here because you dont have annotated mocks. Your mock in line 41 is created instant. The line in setup is necessary if you have instance members which are mocks like

```
@Mock
private MqttBrokersObserver observer;
```

This way `observer` will always be present. You might go this way if you use your mock in more than one test. So ne need to change it here.

htreu replied:
`setUp` is obsolete then, too.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttBrokerConnectionTests.java:
```diff
@@ -28,125 +29,77 @@
-
     @Before
     public void setUp() {
         MockitoAnnotations.initMocks(this);
```
see my comment in MqttServiceTests line 36.

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttBrokerConnectionTests.java:
```diff
@@ -28,125 +29,77 @@
-            fail("addConsumer failed");
-        }
+        MqttMessageSubscriber subscriber = mock(MqttMessageSubscriber.class);
+        when(subscriber.getTopic()).thenAnswer(i -> "topic");
```
Nice! alternative (no need for you to change it here):

`when(subscriber.getTopic()).thenReturn("topic");`

htreu commented on bundles/io/org.eclipse.smarthome.io.transport.mqtt.test/src/test/java/org/eclipse/smarthome/io/transport/mqtt/test/MqttServiceTests.java:
```diff
@@ -19,75 +21,60 @@
+        MqttBrokerConnection connection = new MqttBrokerConnection("name", "tcp://123.123.123.123", false);
         assertTrue(service.addBrokerConnection(connection));
-        assertTrue(brokerExisting);
+        verify(observer).brokerAdded(connection);
```
nice!

davidgraeff commented:
I changed all occurrences of TreeMap to HashMap and removed the not used setups of the test suits.

sjka commented:
Thanks for this contribution! It's a good work for the better of the MQTT transport.
As it exceed 1000 lines of code we unfortunately have to run it through the Eclipse IP process.  This usually takes just a few days, no too bad. The good news for you is that you normally don't need to do anything for it 😄 

So please do not change this PR anymore. If anything would be needed to be changed, then it needs to go through a separate PR. 

For the record: I have created CQ 13824.

sjka commented:
Good news, the CQ was approved!

sjka commented:
Thanks!
